src/server/utils.js

```diff
@@ -0,0 +1,51 @@
+import { Access as AccessCapabilities, Space as SpaceCapabilities } from '@web3-storage/capabilities'
+import { InvalidDelegation } from '../middleware/withDelegationsStorage.js'
+
+/**
+ * Checks if the space/content/serve/* delegation is for the gateway and it is not expired.
+ *
+ * @param {import('@ucanto/interface').Signer} gatewayIdentity - The signer of the gateway identity
+ * @param {import('@ucanto/interface').InferInvokedCapability<typeof AccessCapabilities.delegate>} capability - The capability to validate
+ * @param {import('@ucanto/interface').Proof[]} proofs - The proofs to validate
+ */
+export const extractContentServeDelegation = (gatewayIdentity, capability, proofs) => {
+  const nbDelegations = new Set(Object.values(capability.nb.delegations))
+  if (nbDelegations.size !== 1) {
+    return { error: new InvalidDelegation(`nb.delegations has more than one delegation`) }
+  }
+
+  const delegationLink = Array.from(nbDelegations)[0]
+  const proofDelegations = proofs.flatMap((proof) => 'capabilities' in proof ? [proof] : [])
+  const delegationProof = proofDelegations.find((p) =>
+    delegationLink.equals(p.cid)
+  )
+  if (!delegationProof) {
+    return { error: new InvalidDelegation(`delegation not found in proofs: ${delegationLink}`) }
+  }
+
+  if (delegationProof.audience.did() !== gatewayIdentity.did()) {
+    return {
+      error: new InvalidDelegation(
+        `invalid audience ${delegationProof.audience.did()} does not match Gateway DID ${gatewayIdentity.did()}`
+      )
+    }
+  }
+
+  if (delegationProof.expiration && delegationProof.expiration !== Infinity && delegationProof.expiration < Math.floor(Date.now() / 1000)) {
+    return {
+      error: new InvalidDelegation(
+        `delegation expired at ${delegationProof.expiration}`
+      )
+    }
+  }
+
+  if (!delegationProof.capabilities.some((c) => c.can === SpaceCapabilities.contentServe.can)) {
```

Comment:
I don't think this is necessary at the moment, primarily because our client currently only grants the `space/content/serve/*` delegation. In the future, we might consider adding this if we decide to update the client to delegate `transport/http` and `egress/record` independently. For now, we can simplify by deriving these two capabilities from the top-level capability.